src/Security/Security.php

```diff
@@ -206,15 +213,120 @@ class Security extends Controller implements TemplateGlobalProvider
      * @var boolean If set to TRUE or FALSE, {@link database_is_ready()}
      * will always return FALSE. Used for unit testing.
      */
-    static $force_database_is_ready = null;
+    protected static $force_database_is_ready = null;
 
     /**
      * When the database has once been verified as ready, it will not do the
      * checks again.
      *
      * @var bool
      */
-    static $database_is_ready = false;
+    protected static $database_is_ready = false;
+
+    /**
+     * @var array available authenticators
+     */
+    protected static $authenticators = [];
+
+    /**
+     * @return array
+     */
+    public static function getAuthenticators()
+    {
+        return self::$authenticators;
+    }
+
+    /**
+     * @param array|\SilverStripe\Security\Authenticator $authenticators
+     */
+    public static function setAuthenticators(array $authenticators)
+    {
+        self::$authenticators = $authenticators;
+    }
+
+    /**
+     * @var Member Currently logged in user (if available)
+     */
+    protected static $currentUser;
+
+    /**
+     * @inheritdoc
+     */
+    protected function init()
+    {
+        parent::init();
+
+        // Prevent clickjacking, see https://developer.mozilla.org/en-US/docs/HTTP/X-Frame-Options
+        $frameOptions = static::config()->get('frame_options');
+        if ($frameOptions) {
+            $this->getResponse()->addHeader('X-Frame-Options', $frameOptions);
+        }
+
+        // Prevent search engines from indexing the login page
+        $robotsTag = static::config()->get('robots_tag');
+        if ($robotsTag) {
+            $this->getResponse()->addHeader('X-Robots-Tag', $robotsTag);
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function index()
+    {
+        return $this->httpError(404); // no-op
+    }
+
+    /**
+     * Get the selected authenticator for this request
+     *
+     * @param $name string The identifier of the authenticator in your config
+     * @return Authenticator Class name of Authenticator
+     * @throws LogicException
+     */
+    protected function getAuthenticator($name = 'default')
+    {
+        $authenticators = static::$authenticators;
+
+        if (isset($authenticators[$name])) {
+            return $authenticators[$name];
+        }
+
+        throw new LogicException('No valid authenticator found');
+    }
+
+    /**
+     * Get all registered authenticators
+     *
+     * @param int $service The type of service that is requested
+     * @return array Return an array of Authenticator objects
+     */
+    public function getApplicableAuthenticators($service = Authenticator::LOGIN)
+    {
+        $authenticators = static::$authenticators;
+
+        /** @var Authenticator $class */
+        foreach ($authenticators as $name => $class) {
```

Comment:
Re-assignment of array value by-ref makes this code really awkward to understand. If you implement my above suggestion (inject service instances, not register service names) you can simplify this logic greatly.